### ISSUE TITLE
Audible artefacts when resampling (xwax)

### ISSUE THREAD

**Issue by soundsc** [pideck/pideck#40]
Xwax resampling seems to produce artifacts in the audio output. This was pointed out [in 2010 on xwax-devel](https://sourceforge.net/p/xwax/mailman/message/24547261/) and I just noticed the same issue while playing with the Kool Herc release of PiDeck. The artefacts occurred for me while playing a 44.1 kHz source file with the `-r 48000` option, which is currently the default in PiDeck. For me this was clearly audible e.g. in the kick drum intro of the techno track 'Jerome Sydenham - Whitewater in Trackmode' (a WAV file from Beatport). If you wish I can mix down one bar of output with and without resampling for you to prove the point, but due to time constraints I decided not to do it unless requested.

Is there any reason to enable resampling at all, unless the output sound card does not support the sample rate of the source file? If not, it might be better to disable resampling in the `/usr/share/pideck/pideck` script by default.

I will also raise this on xwax-devel since it really appears to be an issue with resampling in xwax.

Thanks for your fantastic work on this project!

**Comment by danielhjames** [comment]
Hi, thanks for the feedback, we are very interested in fixing audio quality issues! I understand that some audio interfaces are natively 48 kHz and use resampling for 44.1 kHz files. Which audio interface are you using please? Are you using the alsa or jackd backend? Do the artefacts go away if you modify the /usr/share/pideck/pideck script to use the -r 44100 option?

**Comment by soundsc** [comment]
This behaviour was observed identically with:

- American Audio Genie II
- Behringer U-Phono UFO202 (with the little [hardware mod](https://www.mixxx.org/forums/viewtopic.php?f=6&t=2438) that prevents the input from being passed directly to the output)

`BACKEND="ALSA"` in the `pideck` script. I have not tested jackd at all yet.

Playback sounds fine if I omit the `-r 48000` command line option in the call to xwax in the `pideck` script, or if I set `-r 44100` (presumably these two alternatives are equivalent, since 44.1 kHz is xwax's default SR).

As an aside, I am happy to provide command line output for these two audio interfaces (from amixer etc.), if that would help with making future releases support them out of the box (I had to tweak the `pideck` script a bit to get going).

**Comment by soundsc** [comment]
After some additional testing I no longer think my initial ideas about the origin of these artefacts were correct.

First, I converted the original 44.1 kHz WAV to 48 kHz using Audacity (and a 48 kHz AIF using Ableton Live) for comparison, keeping the 16 bits per sample of the original. To my surprise, there was no difference in the playback when loading these in PiDeck/xwax, i.e. `xwax -r 48000` reproduced the artefacts for all versions of the file, while the issue disappeared for `xwax -r 44100` for all versions.

I then connected the interfaces to another Debian 8 machine and actually reproduced the artefacts on both while playing through ALSA via VLC (i.e. xwax is not involved at all). The `dmesg` output on connection suggests that both interfaces run on similar hardware ("Burr-Brown from TI   USB Audio CODEC").

I need to understand better what is going on here, and will post back when I have the facts. Feel free to close this issue meanwhile, since the actual underlying problem seems to be different from what the original post and title suggested.

**Comment by danielhjames** [comment]
Thanks for the test results, please keep us posted!

In the git version we do have some auto soundcard detection, which will be in the next release image. Please see https://github.com/pideck/pideck/blob/master/cards_autodetect and https://github.com/pideck/pideck/blob/master/sup_cards.conf

**Comment by soundsc** [comment]
Those artefacts can been reproduced by _nearest-neighbour_ interpolation from 44.1 kHz to 48 kHz...

**Comment by danielhjames** [comment]
Thanks, I think we should use a variable in the pideck script for sample rate and test the effect on performance.

**Comment by soundsc** [comment]
Yes, that seems like a good idea. For supported cards, perhaps the recommended sample rate could be added as a third argument in `sup_cards.conf` so that it's set by default? After a month of rotation, I still recommend `-r 44100` for the Genie II and UFO202 cards. It is worth mentioning that the audibility of these artefacts is highly dependent on the audio material playing as well as the relative pitch, and non-technical users would normally have no way of knowing why they are there. So having well-chosen default values ensures a good out-of-the-box experience for users of these cards.

Nearest-neighbour shouldn't really be happening. If I'm allowed to speculate again, I suspect it has to do with the way ALSA interacts with some cards, as I have observed it with another card (MindPrint Trio) which as far as I know doesn't have the TI Burr-Brown chip, and without xwax in the signal chain. So it's not (just) these two cards that behave strangely, nor is it purely an xwax issue. Well, that's clearly beyond the scope of this project, and it's good to know that setting the the right sample rate works well in practice.

**Comment by danielhjames** [comment]
Could this be the fix? 
https://github.com/xwax/xwax/commit/33bffba2dc807e7788e4e300641692aaa0ae965e

**Comment by obbardc** [comment]
Think this could be classic symptoms of realtime priority issues.

**Comment by danielhjames** [comment]
Hi @soundsc I have added a samplerate option in https://github.com/pideck/pideck/commit/0f5a2ce0e8c9e48ddc4a3d2d829d9f4059c455a7#diff-2913c7e790ccc21cb2144d3e76cc81c5

On the Citronic AC1-USB which is very similar to the Audio Genie II in appearance, the 44100 option sounds worse to me, with far more noticeable artefacts on the v2 build from http://64studio.net/~chris/pideck/
